### PR TITLE
Use Parser for cast kernel (#4512)

### LINKED ISSUE #4512: Cannot cast string '2021-01-02' to value of Date64 type
**Describe the bug**
Cannot cast string '2021-01-02' to value of Date64 type

I can cast it to `Date32` type for some reason

**To Reproduce**
This is using `datafusion-cli`

```
❯ select arrow_cast('2021-1-2', 'Date64');
Optimizer rule 'simplify_expressions' failed
caused by
Arrow error: Cast error: Cannot cast string '2021-1-2' to value of Date64 type
❯ select arrow_cast('2021-1-2', 'Date32');
+------------------+
| Utf8("2021-1-2") |
+------------------+
| 2021-01-02       |
+------------------+
1 row in set. Query took 0.000 seconds.
```

I found a workaround to cast the string first to `Date32` and then `Date64`
```
❯ select arrow_cast(arrow_cast('2021-1-2', 'Date32'), 'Date64');
+---------------------+
| Utf8("2021-1-2")    |
+---------------------+
| 2021-01-02T00:00:00 |
+---------------------+
1 row in set. Query took 0.000 seconds.
```

**Expected behavior**
I expect to be able to parse '2021-01-02'  successfully as Date64

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #4512

# Rationale for this change
 

Duplicating the parsing logic leads to inconsistencies and is unnecessary.

I still need to benchmark this, but it should also be faster.

# What changes are included in this PR?

# Are there any user-facing changes?

Dates without leading zeros are not supported, as per https://github.com/apache/arrow-rs/issues/3969

### CODE REVIEW

**Comment by tustvold** [comment]
No major regressions and a nice speed bump for date parsing

```
cast utf8 to date32 512 time:   [9.4216 µs 9.4270 µs 9.4332 µs]
                        change: [-77.107% -77.045% -76.983%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  2 (2.00%) high mild
  6 (6.00%) high severe

cast utf8 to date64 512 time:   [42.573 µs 42.581 µs 42.591 µs]
                        change: [-37.832% -37.753% -37.612%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  5 (5.00%) high mild
  2 (2.00%) high severe
```

File: arrow-cast/src/cast.rs [hunk 1]
```diff
@@ -7823,7 +7420,7 @@ mod tests {
 
         let a = StringArray::from(vec![
             "2000-01-01",          // valid date with leading 0s
-            "2000-2-2",            // valid date without leading 0s
+            "2000-2-2",            // invalid date without leading 0s
```
Comment (alamb) [inline arrow-cast/src/cast.rs:7423]: I think we should continue to treat such strings as valid dates as I think it is very common in the real world

For example, postgres recognizes it as a date as well

```
postgres=# select '2000-2-2'::date;
    date
------------
 2000-02-02
(1 row)
```

Not that postgres is the be-all end/all data system but I think it is illustrative that that format is not some fringe case.

File: arrow-cast/src/cast.rs [hunk 2]
```diff
@@ -7841,17 +7438,12 @@ mod tests {
...
-        )
-        .num_days() as i32;
-        assert!(c.is_valid(1)); // "2000-2-2"
-        assert_eq!(date_value, c.value(1));
+        assert!(!c.is_valid(1)); // "2000-2-2"
 
         // test invalid inputs
         assert!(!c.is_valid(2)); // "2000-00-00"
-        assert!(!c.is_valid(3)); // "2000-01-01T12:00:00"
+        assert!(c.is_valid(3)); // "2000-01-01T12:00:00"
```
Comment (alamb) [inline arrow-cast/src/cast.rs:7445]: While surprising that this is a supported "Date" I understanding why

**Review by alamb — COMMENTED** [review]
Making thing go faster by deleting code. What is not to love ❤️  -- really nice work @tustvold. 

I think we need to fix the regression in parsing dates like `2022-2-2` but then it will be ready to go

BTW While reading the code I noticed that the comments on `Parser` need to be updated (as they are now also called in the cast kernel)

```rust
/// Specialized parsing implementations
/// used by csv and json reader
pub trait Parser: ArrowPrimitiveType {

```

**Comment by tustvold** [comment]
> I think we need to fix the regression in parsing dates like 2022-2-2 but then it will be ready to go

So I have code that does this but it represents a 10% performance regression and also seems a touch inconsistent, as we don't permit unpadded timestamps (as they aren't valid according to RFC3339). How strongly do you feel about this?

**Comment by alamb** [comment]
> So I have code that does this but it represents a 10% performance regression and also seems a touch inconsistent, as we don't permit unpadded timestamps (as they aren't valid according to RFC3339). How strongly do you feel about this?

I feel quite strongly -- I think this would represent a significant functionality regression

**Comment by tustvold** [comment]
I recommend viewing with the whitespace blind diff, for some reason parse.rs had CLRF line endings - https://github.com/apache/arrow-rs/pull/4513/commits/c1aa19a90734ef6cd57744222dc06fdf273717c8?w=1

> I feel quite strongly

I ended up implementing a compromise that special cases dates without timestamps and supports missing whitespace. This avoids regressing the timestamp parsing performance

File: arrow-cast/src/cast.rs [hunk 3]
```diff
@@ -7874,13 +7471,13 @@ mod tests {
         assert_eq!(946728000000, c.value(0));
         assert!(c.is_valid(1)); // "2020-12-15T12:34:56"
         assert_eq!(1608035696000, c.value(1));
-        assert!(c.is_valid(2)); // "2020-2-2T12:34:56"
-        assert_eq!(1580646896000, c.value(2));
+        assert!(!c.is_valid(2)); // "2020-2-2T12:34:56"
```
Comment (alamb) [inline arrow-cast/src/cast.rs:7474]: I think this is OK, as I don't think that format is common for "dates"

File: arrow-cast/src/parse.rs [hunk 4]
```diff
@@ -1,2232 +1,2309 @@
...
+
+        // Parsed as offset from local
+        let date = string_to_datetime(&local, "2020-09-08 13:42:29").unwrap();
+        assert_eq!(dt, date.naive_local());
+        assert_ne!(dt, date.naive_utc());
+    }
+
+    #[test]
+    fn parse_date32() {
+        let cases = [
```
Comment (alamb) [inline arrow-cast/src/parse.rs:1532]: I recommend some error cases here as well

```
(empty string)
80-01-01
342
Foo
2020-09-08-03
2020--04-03
2020--
```

**Review by alamb — APPROVED** [review]
Thank you @tustvold 

Not sure why this PR renderes so large on Github (maybe due to line ending changes)

Whitespace blind diff makes it look much better: https://github.com/apache/arrow-rs/pull/4513/files?w=1

File: arrow-cast/src/cast.rs (hunk 3 above)
Comment (tustvold) [reply arrow-cast/src/cast.rs:7474]: Yeah, it is also not correct according to RFC3339

File: arrow-cast/src/parse.rs [hunk 5]
```diff
@@ -648,7 +648,7 @@ impl Parser for Date32Type {
 
 impl Parser for Date64Type {
     fn parse(string: &str) -> Option<i64> {
-        if string.len() < 10 {
+        if string.len() <= 10 {
```
Comment (alamb) [inline arrow-cast/src/parse.rs:651]: My error cases maybe caught a bug!

**Comment by alamb** [comment]
I updated this PR description to cross out

>  Dates without leading zeros are not supported, as per https://github.com/apache/arrow-rs/issues/3969

Without that this PR may not have any actual API changes

File: arrow-cast/src/parse.rs (hunk 5 above)
Comment (tustvold) [reply arrow-cast/src/parse.rs:651]: Afraid not, this just would impact perf

Comment (alamb) [reply arrow-cast/src/parse.rs:651]: Some day

**Pull request merged**